### PR TITLE
Key-value aggregations and lookups

### LINKED ISSUE #59: Key value json output and key value loader to emulate left joins
Allowing to load dictionaries (e.g. key/value from json files), and in the query do dictionaries look-ups. 
This would cover a kind of LEFT JOIN with an equality condition. e.g. 
```
SELECT adict[id] ... 
```
If `id` does not exist in adict , a `NULL`  value should be returned.

### PR DESCRIPTION
**Pull request by dcmoura**
This PR closes #59.

First, it introduces `dict_agg` for key-value aggregations. Example:

```
$ cat codes.csv
MCC,MCC (int),MNC,MNC (int),ISO,Country,Country Code,Network
289,649,88,2191,ge,Abkhazia,7,A-Mobile
289,649,68,1679,ge,Abkhazia,7,A-Mobile
289,649,67,1663,ge,Abkhazia,7,Aquafon
412,1042,01,31,af,Afghanistan,93,AWCC
412,1042,50,1295,af,Afghanistan,93,Etisalat
412,1042,30,783,af,Afghanistan,93,Etisalat
452,1106,04,79,vn,Vietnam,84,Viettel
452,1106,02,47,vn,Vietnam,84,VinaPhone
543,1347,299,665,wf,Wallis and Futuna,,Failed Calls
543,1347,01,31,wf,Wallis and Futuna,,Manuia
421,1057,999,2457,ye,Yemen,967,Fix Line
421,1057,04,79,ye,Yemen,967,HITS/Y Unitel
421,1057,01,31,ye,Yemen,967,Sabaphone
421,1057,03,63,ye,Yemen,967,Yemen Mob. CDMA
645,1605,01,31,zm,Zambia,260,Airtel
645,1605,299,665,zm,Zambia,260,Failed Calls

$ spyql "
    SELECT dict_agg(MCC, Country) AS json 
    FROM csv TO json
  " < codes.csv > code_country.json

$ jq . code_country.json                                        
{
  "289": "Abkhazia",
  "412": "Afghanistan",
  "452": "Vietnam",
  "543": "Wallis and Futuna",
  "421": "Yemen",
  "645": "Zambia"
}
```

Second, it introduces the ~~`kv` function~~ `-J` option to allow loading JSON objects for key-value lookups, emulating left equi joins. Example:

```sh
$ cat towers_mini.csv 
radio,mcc,net,area,cell,unit,lon,lat,range,samples,changeable,created,updated,averageSignal
UMTS,262,2,776,165186,0,13.329163,52.46521,1000,2,1,1286645042,1295769656,0
GSM,262,3,1075,5651,0,13.329163,52.46521,1000,2,1,1286645042,1295769656,0
UMTS,262,2,801,165123,0,13.295516967773,52.494735717773,1000,1,1,1286864565,1286864565,0
UMTS,262,2,801,165121,0,13.301697,52.499886,1000,4,1,1286884439,1297735807,0
GSM,624,2,6,13883,0,9.704361,4.063911,1588,8,1,1352031870,1374212876,0
GSM,624,2,6,34381,0,9.709009,4.066368,2484,13,1,1352031870,1380389330,0
GSM,452,1,10068,5293,0,105.8031463623,20.959854125977,1000,2,1,1459692342,1488347104,0
GSM,645,3,130,3282,0,28.070755004883,-14.902267456055,1000,1,1,1459743588,1459743588,0
GSM,645,3,1,32221,0,28.252716,-15.439224,1000,2,1,1369353852,1369353852,0
GSM,645,3,1,33932,0,28.255325,-15.436752,1000,5,1,1369353888,1369353888,0
GSM,645,3,1,31662,0,28.258072,-15.434555,1000,5,1,1369353960,1369353960,0

$ spyql -Jcountries=code_country.json "SELECT mcc, countries[mcc] AS country, count_agg(1) AS n_recs FROM csv GROUP BY 1, 2 ORDER BY 1 TO pretty" < towers_mini.csv
  mcc  country      n_recs
-----  ---------  --------
  262                    4
  452  Vietnam           1
  624                    2
  645  Zambia            4
```

### CODE REVIEW

**Comment by recharte** [comment]
Well done Daniel :clap:
The functionality introduced with the `kv()` function is awesome however I'm still unsure about the syntax used.
In my mind, the most intuitive way would be to separate the value access in the query from the file loading. What do you think?

**Comment by dcmoura** [comment]
> Well done Daniel 👏
> The functionality introduced with the `kv()` function is awesome however I'm still unsure about the syntax used.
> In my mind, the most intuitive way would be to separate the value access in the query from the file loading. What do you think?

@recharte: you are touching (IMO) the most critical point of this functionality. Having a function like `kv(file, key)` or `kv(file)[key]` is the most straightforward approach as it does not change the query structure. Let me put some alternatives and let me know what you think:

1. `spyql "SELECT kv('file.json', id) FROM csv"`: current approach
2. `spyql "SELECT kv('file.json')[id] FROM csv"`: possible with current approach 
3. `spyql "LOAD kv('file.json') AS my_map SELECT my_map[id] FROM csv"`: introduces the LOAD statement that can be used to load datasets into memory
4. `spyql "INIT my_map=kv('file.json') SELECT my_map[id] FROM csv"`: introduces the INIT statement (a python expression that runs once and before the query) that can be used to load datasets into memory, among other things like definition of constants and calling init functions from other python packages (e.g. initialising random seeds)
5. `spyql -Kmy_map=file.json "SELECT my_map[id] FROM csv"`: introduces the -K command-line argument to load key-value datasets into memory. No changes to the query structure, but you need context to understand it
6. `spyql "SELECT my_map[id] FROM csv, kv('file.json') AS my_map"`: introduces multiple data sources on FROM, which surfaces a lot of questions and decisions about JOINs, data source alias, etc (we can think on variations of this alternative using JOINs)

One more thing: if you are using spyql as a lib in a python script, you can already do this, it's just making use of a dict variable that is available, which goes in the lines of option 5. e.g. https://github.com/dcmoura/spyql/blob/3ddfe45ce54c777e3c59ec7097e5ff109f7566de/tests/files_lib_test.py#L171
https://github.com/dcmoura/spyql/blob/3ddfe45ce54c777e3c59ec7097e5ff109f7566de/tests/files_lib_test.py#L172


Summing up, some decisions:
- Do we want the file loading to be part of the query or shall we leave it outside (like in 5. or currently when using the lib)?
- If inside the query, do we want a pythonic file loading (4.) a SQLish (3.) or SQL (6.)?
- Do we want to keep FROM with a single datasource, or do we want to open to multiple data sources (this is more philosophical)?

**Comment by recharte** [comment]
First, let me share my opinion on the more philosophical question.
> Do we want to keep FROM with a single datasource, or do we want to open to multiple data sources (this is more philosophical)?

Option (6) is interesting as it opens many new possibilities, however I have mixed feelings about it.
As it stands, the FROM expression is already slightly confusing because it can be used for different purposes:
1. Specifing the data type/format received in stdin (`csv`, `spy`, `text`, `json`)
2. Generating data, e.g. https://github.com/dcmoura/spyql#python-iteratorlistcomprehension-to-json
3. Specifing the data type/format received in stdin and selecting a library for processing `orjson`

Options (3) and (5) were the ones that came through my mind when I was reviewing this PR.

> Do we want the file loading to be part of the query or shall we leave it outside (like in 5. or currently when using the lib)?

I do like the simplicity it brings to the query itself and the consistency with the different ways one can use SpyQL (CLI or lib). It brings a good separation of concerns between the inputs and the processing.

> If inside the query, do we want a pythonic file loading (4.) a SQLish (3.) or SQL (6.)?

Here, I don't have good arguments for either. The one that feels more natural to me is (3), though.

**Comment by dcmoura** [comment]
> > Do we want the file loading to be part of the query or shall we leave it outside (like in 5. or currently when using the lib)?
> 
> 
> 
> I do like the simplicity it brings to the query itself and the consistency with the different ways one can use SpyQL (CLI or lib). It brings a good separation of concerns between the inputs and the processing.

Yes, I agree. Option 5 is both consistent and simple to implement. Let's go for it, thanks!

**Comment by dcmoura** [comment]
> > > Do we want the file loading to be part of the query or shall we leave it outside (like in 5. or currently when using the lib)?
> > 
> > 
> > I do like the simplicity it brings to the query itself and the consistency with the different ways one can use SpyQL (CLI or lib). It brings a good separation of concerns between the inputs and the processing.
> 
> Yes, I agree. Option 5 is both consistent and simple to implement. Let's go for it, thanks!

@recharte: I have updated the implementation as discussed. The example in the PR was also updated. Tell me what you think :-)

**Review by ricardocchaves — APPROVED** [review]
Have been following your discussion from the sidelines, I agree with this approach. Great work, this is a cool and useful feature. Code itself looks good to me, no blockers.

However, I have two comments:
1. missing `dict_agg` documentation: I'm not sure [here](https://github.com/dcmoura/spyql/blob/master/spyql/cli.py#L75) is the right place to do it, or if an example in README would be more adequate. The example you wrote in this PR's description was super clear, but I'm afraid of it being the only doc/mention there is on it (other than the code itself);
2. I noticed that the README is getting a bit large, and due to not having seen it for a while, felt some difficulties finding the right section I was looking for. I feel like a table of contents would be beneficial. What do you think @recharte @dcmoura? (I can open a PR for that if you don't want to do it in this one) Not sure there is an automatic way to do a ToC in github markdown, so perhaps it would have to be manual, but I don't think new sections are added often enough for it to be cumbersome.

**Comment by dcmoura** [comment]
> Have been following your discussion from the sidelines, I agree with this approach. Great work, this is a cool and useful feature. Code itself looks good to me, no blockers.
> 
> However, I have two comments:
> 
> 1. missing `dict_agg` documentation: I'm not sure [here](https://github.com/dcmoura/spyql/blob/master/spyql/cli.py#L75) is the right place to do it, or if an example in README would be more adequate. The example you wrote in this PR's description was super clear, but I'm afraid of it being the only doc/mention there is on it (other than the code itself);
> 2. I noticed that the README is getting a bit large, and due to not having seen it for a while, felt some difficulties finding the right section I was looking for. I feel like a table of contents would be beneficial. What do you think @recharte @dcmoura? (I can open a PR for that if you don't want to do it in this one) Not sure there is an automatic way to do a ToC in github markdown, so perhaps it would have to be manual, but I don't think new sections are added often enough for it to be cumbersome.

**Comment by dcmoura** [comment]
Closing this PR. 
Merge was done manually into master. 
Feel free to challenge the implementation :-)

**Review by recharte — COMMENTED** [review]
changes look good :+1:
